lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp

```diff
@@ -414,19 +343,243 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
+
+    // Unmunge output values
+    for (const auto &it : llvm::enumerate(outVals)) {
+      if (isInt1)
+        outVals[it.index()] =
+            icmp_ne(it.value(), rewriter.create<LLVM::ConstantOp>(
+                                    loc, i8_ty, rewriter.getI8IntegerAttr(0)));
+      else if (isPtr)
+        outVals[it.index()] = inttoptr(llvmElemTyOrig, it.value());
+    }
+
+    Value result = packLLElements(loc, getTypeConverter(), outVals, rewriter,
+                                  op.getType());
+    rewriter.replaceOp(op, result);
+    return success();
   }
 
-  LogicalResult
-  transferWithinBlockGroup(const LinearLayout &conversion, ConvertLayoutOp op,
-                           OpAdaptor adaptor,
-                           ConversionPatternRewriter &rewriter) const {
-    // TODO(jlebar): Implement me.
-    return failure();
+  SmallVector<Value> transferWithinBlockOrGroupImpl(
+      ArrayRef<Value> inVals, const LinearLayout &conversion,
+      ConvertLayoutOp op, const LinearLayout &srcLayout,
+      const LinearLayout &dstLayout, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const {
+    MLIRContext *ctx = op.getContext();
+    auto loc = op.getLoc();
+
+    auto sharedPtrTy = ptr_ty(ctx, /*addressSpace=*/3);
+    bool isCrossCTA = isCrossCTAConversion(conversion);
+
+    StringAttr kRegister = str_attr("register");
+    StringAttr kLane = str_attr("lane");
+    StringAttr kWarp = str_attr("warp");
+    StringAttr kBlock = str_attr("block");
+    StringAttr kOffset = str_attr("offset");
+    StringAttr kIteration = str_attr("iteration");
+
+    Value threadId = getThreadId(rewriter, loc);
+    Value threadsPerWarp = i32_val(srcLayout.getInDimSize(kLane));
+    Value laneId = urem(threadId, threadsPerWarp);
+    Value warpId = udiv(threadId, threadsPerWarp);
+
+    auto scratchConfig = ScratchConfig::get(op);
+    auto tensorShape = op.getType().getShape();
+    // Convert int64_t to unsigned
+    SmallVector<unsigned> tensorShapeUnsigned(tensorShape.begin(),
+                                              tensorShape.end());
+    // Input dims: [offset, iteration, block]
+    // Output dims: dimN-1, dimN-2, ..., dim0, where N is obtained from repShape
+    LinearLayout sharedLayout =
+        mlir::triton::gpu::chooseShemLayoutForRegToRegConversion(
+            ctx, tensorShapeUnsigned, scratchConfig.repShape);
+
+    // Layout for the store from registers to shared memory.
+    //
+    // Note: If two threads in the same warp write to the same shmem offset, the
+    // hardware resolves that without a stall or a bank conflict.  Therefore we
+    // don't need to avoid duplicate writes.
+    LinearLayout shmemStoreLayout = srcLayout.invertAndCompose(sharedLayout);
+    const int shmemAllocatedNumElems =
+        getTotalSize<unsigned>(scratchConfig.paddedRepShape);
+    assert(shmemStoreLayout.getOutDimSize(kOffset) <= shmemAllocatedNumElems);
```

Comment:
I wonder if this should be == in your current implementation?

Reply:
It will still be <= because we may have padded elements not considered by kOffset